### PR TITLE
Transitively simple objects

### PR DESCRIPTION
**Pull request by hermanventer**
Release note: Provides a way to make every unknown property value from a simple object be a simple object. Fixes issue: #1410

### CODE REVIEW

**Comment by NTillmann** [comment]
So I put out #1449. It does fix a basic ordering problem (new test cases included).
However, then I tried to combine my changes there with your changes here, and still got wrongly ordered code.

I dug a bit into it. Issues are:
- Generates are linked up in two directions: Each generator has a list of `dependencies`, i.e. generates that will be recursively visited. But each generator also has a pointer back to its owner (the `.parent` property). This backpointer chain is used to compute the "common ancestor". However, this chain seems largely broken: Functions like `joinGenerators` properly set the forward `dependencies`, but the `.parent` link seems generally wrong (always pointing to the root generator?). This wrong linkage can be inspected when running with the --debugScopes option.
- There are generator entries that define the deep property accesses, e.g. `var _$1 = _a.y;`. There, `_$1` is an `AbstractValue` (the only temporal thing we have). However, supposedly subsequent *accesses* are in fact instances of `ObjectValue` (the template object?) that happen to also have  intrinsic names such as `_$1`. And thus the ordering doesn't get respected.

**Comment by NTillmann** [comment]
Two more thoughts for the "transitive simple object" feature, and how it will be used:
- The incoming object isn't just known to be transitively simple, we can also assume that it is immutable.
- Thus, we don't need temporal values at all. If the build nodes are safe, e.g. always guard a member access by a null check, plus the previous immutability assumption, and non-temporal abstract values should do.

**Comment by hermanventer** [comment]
Instead of assuming that is could we just look too see if it is frozen?

**Comment by hermanventer** [comment]
Actually, there is no simple way to transitively freeze objects, so the _simplicityIsTransitive flag is a better thing to test. Done.

**Comment by hermanventer** [comment]
Hmm. Treating transitive simple objects as readonly breaks a test. Reverting that for now.

File: src/values/ObjectValue.js [hunk 1]
```diff
@@ -244,7 +247,15 @@ export default class ObjectValue extends ConcreteValue {
...
-  isTemplate: void | true;
+  // If true, then unknown properties should return transitively simple abstract object values
+  _simplicityIsTransitive: BooleanValue;
+
+  // Specifies whether the object is a template that needs to be created in a scope
+  _isScopedTemplate: void | true;
+
+  // The abstract object for which this object is the template.
+  // Use this instead of the object itself when deriving temporal values for object properties.
+  _abstractObject: void | AbstractObjectValue;
```
Comment (NTillmann) [inline src/values/ObjectValue.js]: maybe call it `_templateOf`

File: src/values/ObjectValue.js [hunk 2]
```diff
@@ -244,7 +247,15 @@ export default class ObjectValue extends ConcreteValue {
   // to return AbstractValue for unknown properties.
   _isSimple: BooleanValue;
 
-  isTemplate: void | true;
+  // If true, then unknown properties should return transitively simple abstract object values
+  _simplicityIsTransitive: BooleanValue;
```
Comment (NTillmann) [inline src/values/ObjectValue.js]: `| AbstractValue` --- let's do this new one right...

File: src/serializer/ResidualHeapVisitor.js [hunk 3]
```diff
@@ -755,8 +755,8 @@ export class ResidualHeapVisitor {
       if (this.preProcessValue(val)) this.visitAbstractValue(val);
     } else if (val.isIntrinsic()) {
       // All intrinsic values exist from the beginning of time...
-      // ...except for a few that come into existence as templates for abstract objects (TODO #882).
-      if (val.isTemplate) this.preProcessValue(val);
+      // ...except for a few that come into existence as templates for abstract objects via executable code (TODO #882).
```
Comment (NTillmann) [inline src/serializer/ResidualHeapVisitor.js]: Note that #882 is already closed --- time to remove the (TODO ...) text?

File: src/values/ObjectValue.js [hunk 4]
```diff
@@ -297,8 +308,12 @@ export default class ObjectValue extends ConcreteValue {
     this._isPartial = this.$Realm.intrinsics.true;
   }
 
-  makeSimple(): void {
+  makeSimple(option?: any): void {
```
Comment (NTillmann) [inline src/values/ObjectValue.js]: Avoid : any type.

File: src/values/ObjectValue.js (hunk 2 above)
Comment (hermanventer) [reply src/values/ObjectValue.js]: Sure, but in another pull request.

**Pull request closed without merging**